### PR TITLE
[eslint-plugin-react-hooks] Fix cyclic caching for loops containing a…

### PR DESCRIPTION
**Pull request by M-Izadmehr**
This MR Fixes a bug (#16832) in the `eslint-plugin-react-hooks`, where caching of cycles is performed wrong in some cases.

**Example:** https://codesandbox.io/s/exciting-bhabha-mqj7q

**Description**
If the user uses an `if` statement inside a `for ... of obj` statement, and uses a correct hook afterwards, the eslint shows an *error* that:
`React Hook "${hook}" is called conditionally. React Hooks must be called in the exact same order in every component render.`

**Reports**
I guess that it is already known issue because, inside the [ESLintRulesOfHooks.js - L562](https://github.com/facebook/react/blob/master/packages/eslint-plugin-react-hooks/__tests__/ESLintRulesOfHooks-test.js#L562), it is already known that in some cases caching of cyclic paths in code path, might result in the wrong errors.

Also (#16832) reports this case too.

**Reason**
Eslint uses `CodePathSegment` which is similar to doubly linked list. Based on the `CodePathSegment` AST different types of cyclic segments might happen. In the special case of [ForInStatement](https://eslint.org/docs/developer-guide/code-path-analysis#forinstatement) there are two paths from obj identifier to the next segment. In such a graph, in order to traverse all the paths from start to end, we need to not only remember the current segment, but also the whole path. 

The current code uses a simple caching just by name of segments (and not the list of segment history) as a result, whenever it faces a cycle, it breaks the cycle by returning 0. As a result, there might be several conditions, where the segment might have cyclic and not cyclic paths to end. In the current code, if the first path is cyclic all the non-cyclic ones are thrown away. And depending on the starting point and endpoint of AST, we might have different results.

For example for the demo code, the result of  [RulesOfHooks.js - L404](https://github.com/facebook/react/blob/master/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L291), the value of all `allPathsFromStartToEnd` is calculated as `2` instead of `3`

**Solution**
We need caching of responses and also saving cyclic paths for later logic, In the new code, I added a different caching mechanism, which also remembers the path history too. The effect of the performance is still negligible because even for [the most difficult test](https://github.com/facebook/react/blob/master/packages/eslint-plugin-react-hooks/__tests__/ESLintRulesOfHooks-test.js#L293) including `1099511627776` paths from start to end of the code, it only increases the time about 0.3s on MacBookPro.

### CODE REVIEW

File: packages/eslint-plugin-react-hooks/src/RulesOfHooks.js [hunk 1]
```diff
@@ -114,39 +114,41 @@ export default {
...
-          if (paths === null) {
-            if (cyclic.has(segment.id)) {
-              return 0;
-            } else {
-              cyclic.add(segment.id);
-              for (const prevSegment of segment.prevSegments) {
-                countPathsFromStart(prevSegment);
-              }
-              return 0;
+          let pathList = new Set(pathHistory);
```
Comment (kidkkr) [inline packages/eslint-plugin-react-hooks/src/RulesOfHooks.js]: pathList doesn't seem to be reallocated.

```suggestion
          let pathList = new Set(pathHistory);
```

File: packages/eslint-plugin-react-hooks/src/RulesOfHooks.js [hunk 2]
```diff
@@ -183,43 +185,45 @@ export default {
...
           } else {
             paths = 0;
             for (const nextSegment of segment.nextSegments) {
-              paths += countPathsToEnd(nextSegment);
+              paths += countPathsToEnd(nextSegment, pathList);
             }
           }
-          cache.set(segment.id, paths);
 
+          cache.set(segment.id, paths);
```
Comment (kidkkr) [inline packages/eslint-plugin-react-hooks/src/RulesOfHooks.js]: I think unnecessary diff in here.

Comment (M-Izadmehr) [reply packages/eslint-plugin-react-hooks/src/RulesOfHooks.js]: I am not sure why, but I am not sure why it considers this line as changed:)

**Comment by TrySound** [comment]
@gaearon This problem is still relevant.

**Comment by gaearon** [comment]
Thanks!

**Comment by M-Izadmehr** [comment]
I almost forgot about this :D
